aarch64: Fix alignment fault on z_bss_zero()

carlocaione opened the pull request:
Using newlibc with AArch64 is causing unaligned fault in z_bss_zero()
when the code is run on real hardware (on QEMU the problem is not
reproducible).

The main cause is that the memset() function exported by newlibc is
using 'DC ZVA' to zero out memory.

While this is often a nice optimization, this is causing the issue on
AArch64 because memset() is being used before the MMU is enabled, and
when the MMU is disabled all data accesses will be treated as
Device_nGnRnE.

This is a problem because quoting from the ARM reference manual: "If the
memory region being zeroed is any type of Device memory, then DC ZVA
generates an Alignment fault which is prioritized in the same way as
other alignment faults that are determined by the memory type".

newlibc tries to be a bit smart about this reading the DCZID_EL0
register before deciding whether using 'DC ZVA' or not. While this is a
good idea for code running in EL0, currently the Zephyr kernel is
running in EL1. This means that the value of the DCZID_EL0 register is
actually retrieved from the HCR_EL2.TDZ bit, that is always 0 because
EL2 is not currently supported / enabled. So the 'DC ZVA' instruction is
unconditionally used in the newlibc memset() implementation.

The "standard" solution for this case is usually to use a different
memset routine to be specifically used for two cases: (1) against IO
memory or (2) against normal memory but with MMU disabled (which means
all memory is considered device memory for data accesses).

To fix this issue in Zephyr we avoid calling memset() when clearing the
bss, and instead we use a simple loop to zero the memory region.

pabigot commented on lib/os/mem.c:
```diff
@@ -0,0 +1,35 @@
+	char *s8;
+	int i;
+
+	if (((unsigned long)s & (sizeof(*sl) - 1)) == 0) {
```
If the request is to set a large region where the start address isn't word-aligned the behavior is pessimized.  Is it worth adding a pre-align loop like the post-align loop to deal with this?

carlocaione replied:
We can use whatever memset() we want, I just took one from the linux kernel without much thought TBH. In theory I could have copied the one shipped in the minimal libc that we already ship in `lib/libc/minimal/source/string/string.c`. What about that one?

pabigot replied:
Moving that implementation here, and just calling it from minimal libc, would probably be better.  I'm now concerned about "just took the one from the linux kernel", since that's got licensing/copyright implications.

carlocaione replied:
I know, but there are only a few permutations of a memset() around and in general this is an RFC because I was more concerned with the approach to fix the issue than the actual code :) I'll move that implementation here, no prob.

pabigot reviewed (commented):
I don't see a problem with this; I've run into a similar issue in the past but with copying data to/from device memory.

If it's going to get merged it'd be nice to use some trick so that minimal libc `memset` calls the same function.

npitre commented:
I'm not convinced by this approach. This is adding a somewhat largish 
memset implementation that is going to be used only once.

Instead I'd simply do a plain while() loop in z_arm64_bss_zero(). 
Something like:

```
static void z_arm64_bss_zero(void)
{
	uint64_t *p = (uint64_t *)__bss_start;
	uint64_t *end = (uint64_t *)__bss_end;
	while (p < end) {
		*p++ = 0;
	}
}
```

No need to be fancier than that. There shouldn't be alignment nor 
overlap issues with either boundaries.

carlocaione commented:
> I'm not convinced by this approach. This is adding a somewhat largish memset implementation that is going to be used only once. 

Well, this brings me to my second question really. I have only used `sys_memset()` for `z_bss_zero()` because on AArch64 this is the only place where **currently** `memset()` is used before the MMU is initialized. But in general we have a bunch of other `memset()`s being scattered in the init kernel code. Should we touch all of those as well?

My rationale here to have a specific function here is that we can use this function whenever on AArch64 in the future or maybe on other architectures we see a failing `memset()` during the kernel init.

npitre commented:
We all want to get into a state where the most efficient memset can be 
used as soon as possible.

If we need to have a special unoptimized version in the mean time then 
it has to be a special unoptimized version i.e. with the smallest 
footprint possible. No need for byte paddings since this shouldn't be 
necessary for those few special cases. That boils down to that simple 
loop. And at that point it is probably best to do the loop inline since 
it probably has the same footprint as the equivalent function call.

andrewboie commented:
> this is causing the issue on AArch64 because memset() is being used before the MMU is enabled

Can you turn the MMU on earlier? Or implement a custom bss zero routine?

This really seems an aarch64 problem, would rather see solutions implemented in arch/arm and not in common code.

carlocaione committed:
aarch64: Fix alignment fault on z_bss_zero()

Using newlibc with AArch64 is causing an alignement fault in
z_bss_zero() when the code is run on real hardware (on QEMU the problem
is not reproducible).

The main cause is that the memset() function exported by newlibc is
using 'DC ZVA' to zero out memory.

While this is often a nice optimization, this is causing the issue on
AArch64 because memset() is being used before the MMU is enabled, and
when the MMU is disabled all data accesses will be treated as
Device_nGnRnE.

This is a problem because quoting from the ARM reference manual: "If the
memory region being zeroed is any type of Device memory, then DC ZVA
generates an Alignment fault which is prioritized in the same way as
other alignment faults that are determined by the memory type".

newlibc tries to be a bit smart about this reading the DCZID_EL0
register before deciding whether using 'DC ZVA' or not. While this is a
good idea for code running in EL0, currently the Zephyr kernel is
running in EL1. This means that the value of the DCZID_EL0 register is
actually retrieved from the HCR_EL2.TDZ bit, that is always 0 because
EL2 is not currently supported / enabled. So the 'DC ZVA' instruction is
unconditionally used in the newlibc memset() implementation.

The "standard" solution for this case is usually to use a different
memset routine to be specifically used for two cases: (1) against IO
memory or (2) against normal memory but with MMU disabled (which means
all memory is considered device memory for data accesses).

To fix this issue in Zephyr we avoid calling memset() when clearing the
bss, and instead we use a simple loop to zero the memory region.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

tyrantlizard77 reviewed (approved):
thanks!

povergoing commented:
I have checked it on my platform, and it **works**
